### PR TITLE
add addIceCandidate(null)

### PR DESCRIPTION
**Pull request by fippo**
**Description**
this allows calling addIceCandidate with the null candidate, a relatively recent spec change.

@taylor-b you might know the issue number in chrome
@nils-ohlmeier same for mozilla

**Purpose**
spec-compliant code, today :-)

Detailed comments inline

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -1332,19 +1332,16 @@ test('Basic connection establishment', function(t) {
...
-            // TODO: Decide if we are interested in adding all candidates
-            // as passed tests.
-            tc.pass('addIceCandidate ' + counter++);
-          },
-          function(err) {
-            tc.fail('addIceCandidate ' + err.toString());
-          }
-        );
-      }
+      pc.addIceCandidate(event.candidate,
```
Comment (fippo) [inline test/test.js:1335]: we did not use promises here on purpose, right?

File: src/js/chrome/chrome_shim.js [hunk 2]
```diff
@@ -225,6 +225,14 @@ var chromeShim = {
             };
           });
     }
+
+    // support for addIceCandidate(null)
```
Comment (fippo) [inline src/js/chrome/chrome_shim.js:227]: I decided not to mix this with the promise-ification. Which we might have to remove soon for canary?
Maybe we should separate https://github.com/webrtc/adapter/pull/266/files as well?

File: test/test.js (hunk 1 above)
Comment (KaptenJansson) [inline test/test.js:1335]: I cannot remember if it was on purpose or just out of habit. I assume most are not using promises for this yet?

File: src/js/chrome/chrome_shim.js (hunk 2 above)
Comment (KaptenJansson) [inline src/js/chrome/chrome_shim.js:227]: SGTM on both counts. Separate PR though.

File: src/js/edge/edge_shim.js [hunk 3]
```diff
@@ -905,38 +905,44 @@ var edgeShim = {
...
-        // A dirty hack to make samples work.
-        if (cand.type === 'endOfCandidates') {
-          cand = {};
-        }
-        transceiver.iceTransport.addRemoteCandidate(cand);
+        var transceiver = this.transceivers[mLineIndex];
+        if (transceiver) {
+          var cand = Object.keys(candidate.candidate).length > 0 ?
+              SDPUtils.parseCandidate(candidate.candidate) : {};
+          // Ignore Chrome's invalid candidates since Edge does not like them.
```
Comment (KaptenJansson) [inline src/js/edge/edge_shim.js:926]: Invalid candidates? Should we file a bug?

Comment (fippo) [inline src/js/edge/edge_shim.js:926]: the tcp ones with port 0. Those are gone now I think.

Comment (KaptenJansson) [inline src/js/edge/edge_shim.js:926]: Aha right, yes that is fixed.

**Comment by KaptenJansson** [comment]
LGTM

File: test/test.js (hunk 1 above)
Comment (fippo) [inline test/test.js:1335]: yeah. which makes it important to test the non-promise versions. So lets assume this was a conscious choice

Comment (KaptenJansson) [inline test/test.js:1335]: SGTM

**Comment by nils-ohlmeier** [comment]
No tickets existed yet, so I created https://bugzilla.mozilla.org/show_bug.cgi?id=1273319

**Comment by taylor-b** [comment]
Just a heads up: this may change in the spec at some point. There's an ambiguous situation where if addIceCandidate is called in the middle of an ICE restart, you don't know if it's a post-restart or pre-restart candidate.

The ICE working group decided to trickle the ufrag to solve this ambiguity. But this ambiguity also exists with a trickled "end-of-candidates". So we could end up putting a ufrag on the end-of-candidates indication, meaning it won't be a simple "null" any longer.

See the discussion here: https://github.com/rtcweb-wg/jsep/issues/251#issuecomment-195171292

**Comment by fippo** [comment]
@taylor-b gee. How about checking for !arguments[0].candidate? That seems safe no matter what is going to be the result of the ufrag/pwd discussion.

**Comment by taylor-b** [comment]
I think this is good for now; it's possible the API won't even change. I just wanted to keep you in the loop.

**Comment by fippo** [comment]
sounds good. I think this will allow doing the right thing in the samples -- just forward all the output of onicecandidate to the peer and feed it into addIceCandidate.

**Pull request merged**